### PR TITLE
[Test] Refactor test/test_reductions.py

### PR DESCRIPTION
**Pull request by KarhouTam**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* #185797
* #187926
* #186352
* #186351
* __->__ #185881

cc @mruberry

### CODE REVIEW

File: test/test_reductions.py [hunk 1]
```diff
@@ -3298,427 +3179,85 @@ def test_histc_min_max_corner_cases_cuda(self, device, dtype):
...
-        def reference_histogram(t, bins, bin_range, weights, density, dtype):
-            np_t, np_bins, np_weights = map(to_np, [t, bins, weights])
-            np_hist, np_bin_edges = np.histogram(
-                np_t, np_bins, range=bin_range, weights=np_weights, density=density
-            )
-            return (
-                torch.from_numpy(np_hist).to(dtype),
-                torch.from_numpy(np_bin_edges).to(dtype),
-            )
+# Tests to ensure that reduction functions employing comparison operators are usable when there
```
Comment (can-gaa-hou) [inline test/test_reductions.py]: The indentation seems not ok

**Comment by KarhouTam** [comment]
@claude check if CI failures relate to the changes

**Comment by KarhouTam** [comment]
@jbschlosser @Skylion007 Could you please take a look? Thanks!

File: test/test_reductions.py [hunk 2]
```diff
@@ -101,7 +102,7 @@ def _reduced_shape(shape, empty_dim_as_none=False, dim=None, keepdim=False):
 
     return result
 
-class TestReductions(TestCase):
+class TestReductionsDevice(TestCase):
```
Comment (fffrog) [inline test/test_reductions.py]: Hi @KarhouTam, thanks for the updates.

Renaming TestReductions to TestReductionsDevice would result in massive diffs in this PR. I think it would be better to keep the original class name and instead make `test_scalar_tensor_dim_compiled_mode` and `test_scalar_tensor_as_dim_argument` device-agnostic.

Rather than relying on class names to distinguish roles or responsibilities, we should leverage `hw_classification` to identify the role of each test class. 

This would be a much cleaner approach.

File: test/test_reductions.py
Comment (fffrog) [inline test/test_reductions.py]: Just another thread note:

There are so many `onlyCPU` decorator, and I think most of them can be refactored into device-agnostic.

Granted, we can do this in the follow-up PRs.

Comment (KarhouTam) [reply test/test_reductions.py]: Got it. Already fixed within the latest commit.

File: test/test_reductions.py [hunk 3]
```diff
@@ -3336,6 +2858,744 @@ def test_histc_min_max_corner_cases_cuda(self, device, dtype):
...
+        # mean_cuda is not implemented for ComplexHalf
+        err_msg = "not implemented for 'ComplexHalf'"
+        ctx = self.assertRaisesRegex(
+            RuntimeError, err_msg) if dtype is torch.chalf else contextlib.nullcontext()
+        with ctx:
+            self.assertEqual(torch.mean(t), expected)
+
+
+
+class TestReductionsOnlyCPU(TestCase):
```
Comment (fffrog) [inline test/test_reductions.py]: Thank you very much for grouping these test cases into a separate class.

Just to make sure I understand the intention correctly, does this new class imply that all the test cases in it are truly CPU-only?

In my opinion, many of these test cases could actually be device-agnostic. The @onlyCPU decorator often exists for historical reasons rather than because the test has an inherent CPU dependency.

Because of that, I'm not sure we should simply move every test decorated with @onlyCPU into this new class. Instead, it might make more sense to review them one by one and determine whether each test really depends on CPU or could be enabled on other accelerators as well.

**Comment by fffrog** [comment]
The rest of changes are good to me, thank you.

File: test/test_reductions.py (hunk 3 above)
Comment (KarhouTam) [reply test/test_reductions.py]: Good catch! Will refactor the test class.

File: test/test_reductions.py [hunk 4]
```diff
@@ -3336,99 +2858,775 @@ def test_histc_min_max_corner_cases_cuda(self, device, dtype):
...
+        self.assertEqual(torch.sum(t), expected)
+
+        # mean_cuda is not implemented for ComplexHalf
+        err_msg = "not implemented for 'ComplexHalf'"
+        ctx = self.assertRaisesRegex(
+            RuntimeError, err_msg) if dtype is torch.chalf else contextlib.nullcontext()
+        with ctx:
+            self.assertEqual(torch.mean(t), expected)
+
+    @onlyCPU
```
Comment (fffrog) [inline test/test_reductions.py]: Can this testcase run on CUDA or other acceleartors?

File: test/test_reductions.py
Comment (fffrog) [inline test/test_reductions.py]: Hi @KarhouTam, I think we need to remove this global variable, because this variable can be hidden in the `instantiate_device_type_test` logic.

File: test/test_reductions.py [hunk 5]
```diff
@@ -3628,316 +3822,143 @@ def test_histogram_error_handling(self, device, dtype):
...
-            # Numpy differs, producing uint32 on Windows
-            if expected.dtype in [np.uint64, np.uint32]:
-                exact_dtype = False
+    def test_min_elementwise(self, device):
+        self._testCSelection(torch.min, min)
 
-            msg = ("Failed to produce expected results! Input tensor was"
-                   f" {t}, torch result is {actual}, and reference result is"
-                   f" {expected}.") if t.numel() < 10 else None
+    def test_max_mixed_devices(self, device):
```
Comment (fffrog) [inline test/test_reductions.py]: Taking this testcase as an example:

I think we should migrate it back to `TestReduction` with something changes.

```Python
    @onlyAccelerator
    def test_max_mixed_devices(self, device):
        a = torch.randn(10, "cpu")
        values = torch.randn(10).to(device)
        indices = torch.LongTensor().to(device)
        self.assertRaises(RuntimeError,
                              lambda: torch.max(a, 0, out=(values, indices)))
        self.assertRaises(RuntimeError,
                              lambda: torch.amax(a, 0, out=values))
```

I should specify the device of input tensor, then copy it to the target device.

File: test/test_reductions.py [hunk 6]
```diff
@@ -3628,316 +3822,143 @@ def test_histogram_error_handling(self, device, dtype):
...
+        if torch.accelerator.is_available():
+            values = torch.randn(10).to(device_type)
+            indices = torch.LongTensor().to(device_type)
+            self.assertRaises(RuntimeError,
+                              lambda: torch.max(a, 0, out=(values, indices)))
+            self.assertRaises(RuntimeError,
+                              lambda: torch.amax(a, 0, out=values))
 
-            self.assertEqual(actual, expected, msg, exact_dtype=exact_dtype)
+    def test_min_mixed_devices(self, device):
```
Comment (fffrog) [inline test/test_reductions.py]: ditto

File: test/test_reductions.py [hunk 7]
```diff
@@ -3628,316 +3822,143 @@ def test_histogram_error_handling(self, device, dtype):
...
+    Tests that do not test CPU-specific features were moved to
+    TestReductions with @onlyCPU decorator.
+    """
 
-    @ops(reference_masked_ops)
-    def test_reference_masked(self, device, dtype, op):
-        """Test masked reduction operations on strided-only tensors using
-        numpy reductions as reference.
-        """
+    def test_sum_parallel(self, device):
```
Comment (fffrog) [inline test/test_reductions.py]: I am not sure whether this testcase can be executed well in other accelerator other than CPU.

File: test/test_reductions.py [hunk 8]
```diff
@@ -3628,316 +3822,143 @@ def test_histogram_error_handling(self, device, dtype):
...
-            if takes_dtype:
-                grad2, = torch.autograd.grad([op(x, dim=0, dtype=torch.double)], [x], gi.double())
-            else:
-                grad2, = torch.autograd.grad([op(x.double(), dim=0)], [x], gi.double())
-            self.assertEqual(grad1, grad2)
-            self.assertEqual(grad2.dtype, torch.float)
+    These tests exercise features inherently tied to CPU execution:
+    - CPU parallel reduction (test_sum_parallel)
+    - CPU-only APIs like map2_ (test_max/min_elementwise)
+    - Mixed CPU-accelerator device error paths (test_max/min_mixed_devices)
```
Comment (fffrog) [inline test/test_reductions.py]: Please remove this line, because those both testcases can be migrated into `TestReduction`

File: test/test_reductions.py [hunk 9]
```diff
@@ -3977,6 +3998,7 @@ def foo_compile():
         self.assertEqual(result_eager.shape, torch.Size([2, 2]))
 
 instantiate_device_type_tests(TestReductions, globals(), allow_xpu=True, allow_mps=True)
+instantiate_device_type_tests(TestReductionsOnlyCPU, globals(), only_for="cpu")
```
Comment (fffrog) [inline test/test_reductions.py]: As the concern @albanD have proposed in another PR, I think we can abandon the `instantiate_device_type_tests` for the device-specific test class, and just use the real device such as `cpu` instead of the `device` parameter, then we can rename the `TestReductionOnlyCPU` to `TestReductionCPU`.

File: test/test_reductions.py [hunk 10]
```diff
@@ -3336,99 +2858,774 @@ def test_histc_min_max_corner_cases_cuda(self, device, dtype):
...
+            t = make_tensor(shape, dtype=dtype, device=device)
+            # Attempt to call torch.nanmean and expect a RuntimeError
+            with self.assertRaisesRegex(
+                RuntimeError,
+                r"nanmean\(\): expected input to have floating point or complex dtype but got \w+"
+            ):
+                torch.nanmean(t)
+
+
+    @onlyCPU
```
Comment (fffrog) [inline test/test_reductions.py]: In this PR, there are still many of `@onlyCPU` decorator, I am not sure whether we can remove it to make other accelerator to run those testcases well.

But from the definition of those testcases, I think most of them can work well in other accelerator. Granted, if some testcases can not work well in other accelerator, I think the `TestReductionCPU` should be the right place to hold them.

**Comment by fffrog** [comment]
Additional notes:

This PR contains too many changes, making it difficult to get approved quickly. Therefore, please try to break similar PRs into smaller PRs in the future. Thank you for your support.

File: test/test_reductions.py (hunk 5 above)
Comment (KarhouTam) [reply test/test_reductions.py]: Fixed.

File: test/test_reductions.py (hunk 6 above)
Comment (KarhouTam) [reply test/test_reductions.py]: Fixed.

File: test/test_reductions.py (hunk 7 above)
Comment (KarhouTam) [reply test/test_reductions.py]: We can try it.

File: test/test_reductions.py (hunk 9 above)
Comment (KarhouTam) [reply test/test_reductions.py]: Fixed.

**Comment by KarhouTam** [comment]
Hi @fffrog. Thanks for your detailed suggestions and comments. I've tried re-refactored the whole test file and please re-review it! Many thanks!

File: test/test_reductions.py
Comment (fffrog) [inline test/test_reductions.py]: I think we can use onlyAccelerator here  and skip the specific accelerator that can not work well.

Comment (fffrog) [inline test/test_reductions.py:1192]: All the @onlyNativeDeviceTypes can be replaced with `onlyAcclerator`

**Review by fffrog — APPROVED** [review]
Overall changes looks good to me except some small nits, thank you.

File: test/test_reductions.py [hunk 11]
```diff
@@ -1591,23 +1484,23 @@ def test_prod_bool(self, device):
             expect = np.prod(np.array(val))
             self.assertEqual(result, expect)
 
-    @onlyCPU
+    @onlyAccelerator
```
Comment (albanD) [inline test/test_reductions.py:1500]: THis is suspicious? We went from only cpu to everything except cpu ?

File: test/test_reductions.py [hunk 12]
```diff
@@ -1591,23 +1484,23 @@ def test_prod_bool(self, device):
...
-            indices = torch.LongTensor().to(device_type)
+            values = torch.randn(10).to(device)
+            indices = torch.LongTensor().to(device)
             self.assertRaises(RuntimeError,
                               lambda: torch.max(a, 0, out=(values, indices)))
             self.assertRaises(RuntimeError,
                               lambda: torch.amax(a, 0, out=values))
 
-    @onlyCPU
+    @onlyAccelerator
```
Comment (albanD) [inline test/test_reductions.py:1512]: same

File: test/test_reductions.py [hunk 13]
```diff
@@ -1300,7 +1196,7 @@ def _amax_wrapper(x, dim=None, keepdims=False):
         self._test_minmax_helper(_amin_wrapper, np.amin, device, dtype)
         self._test_minmax_helper(_amax_wrapper, np.amax, device, dtype)
 
-    @onlyNativeDeviceTypes
+    @onlyAccelerator
```
Comment (albanD) [inline test/test_reductions.py]: Doesn't native device type contain CPU, are we loosing coverage here and other places?

**Review by albanD — CHANGES_REQUESTED** [review]
I think there is quite a bit of loss of coverage here.
Don't hesitate to ask claude on this PR as it is pretty good at finding those.

File: test/test_reductions.py (hunk 11 above)
Comment (KarhouTam) [reply test/test_reductions.py:1500]: Actually this test case depends on a global var `device_type` to get the accelerator type before:
```python
device_type = (
    acc.type if (acc := torch.accelerator.current_accelerator(True)) else "cpu"
)

    @onlyCPU
    def test_max_mixed_devices(self, device):
        a = torch.randn(10, device) # always "cpu"
        if torch.accelerator.is_available():
            values = torch.randn(10).to(device_type) # cuda, xpu, ...
            indices = torch.LongTensor().to(device_type) # cuda, xpu, ...
            self.assertRaises(RuntimeError,
                              lambda: torch.max(a, 0, out=(values, indices)))
            self.assertRaises(RuntimeError,
                              lambda: torch.amax(a, 0, out=values))
```

I have refactored the code by removing the device_type global var, replacing `onlyCPU` with `onlyAccelerator`, and removing the `device` for tensor `a`. These changes maintain the original test coverage:

```python
    @onlyAccelerator
    def test_max_mixed_devices(self, device):
        a = torch.randn(10)
        if torch.accelerator.is_available():
            values = torch.randn(10).to(device)
            indices = torch.LongTensor().to(device)
            self.assertRaises(RuntimeError,
                              lambda: torch.max(a, 0, out=(values, indices)))
            self.assertRaises(RuntimeError,
                              lambda: torch.amax(a, 0, out=values))
```

File: test/test_reductions.py (hunk 12 above)
Comment (KarhouTam) [reply test/test_reductions.py:1512]: ditto

File: test/test_reductions.py (hunk 13 above)
Comment (KarhouTam) [reply test/test_reductions.py]: Good catch. Sorry about that.

**Comment by KarhouTam** [comment]
@claude Analyze the test coverage change

**Comment by KarhouTam** [comment]
@claude how does that look now?

**Comment by KarhouTam** [comment]
Hi @albanD . I've fixed this PR according to your comments. PTAL, thanks!

**Review by albanD — APPROVED** [review]
Awesome. Thanks for carefully tackling this.
